### PR TITLE
glTF2 fallback load PNG and JPG

### PR DESCRIPTION
**Pull request by fire**
Be able to load gltf2 with mismatched mime types.

https://github.com/V-Sekai/godot-blue-steel-wolf/blob/main/mozilla_hubs_importer/art/festival/CC-BY%20Mozilla%20Hubs%20Outdoor%20Festival.glb

image

image

### CODE REVIEW

**Comment by akien-mga** [comment]
Does this respect the spec? IIRC I took good care to do exactly what the spec says.

See: https://github.com/KhronosGroup/glTF/blob/master/specification/2.0/README.md#images

> Implementation Note: When image data is provided by uri and mimeType is defined, client implementations should prefer JSON-defined MIME Type over one provided by transport layer.

Though it says "prefer", which is still what we do here. I suppose it doesn't hurt to add a fallback for bogus assets.

**Comment by fire** [comment]
That's my interpretation too, I wanted to also do a generic image load at the end, but that's a proposal enhancement.

**Comment by Calinou** [comment]
Maybe add a `print_verbose()` line to state that non-spec-compliant PNG loading and JPEG loading is attempted.

**Comment by fire** [comment]
Can you propose some wording?

**Comment by Calinou** [comment]
> Can you propose some wording?

For PNG:

`glTF: No MIME type specified for Base64-encoded glTF image. Attempting to decode the image as PNG.`

For JPEG:

`glTF: No MIME type specified for Base64-encoded glTF image. Decoding as PNG failed, attempting to decode the image as JPEG.`

**Comment by akien-mga** [comment]
> > Can you propose some wording?
> 
> For PNG:
> 
> `glTF: No MIME type specified for Base64-encoded glTF image. Attempting to decode the image as PNG.`
> 
> For JPEG:
> 
> `glTF: No MIME type specified for Base64-encoded glTF image. Decoding as PNG failed, attempting to decode the image as JPEG.`

IIRC it shouldn't be possible to be in this code chunk without a MIME type defined. So the actual case might be that it's an unsupported MIME type.

I'll have a look.

**Comment by akien-mga** [comment]
So https://github.com/V-Sekai/godot-blue-steel-wolf/blob/main/mozilla_hubs_importer/art/festival/CC-BY%20Mozilla%20Hubs%20Outdoor%20Festival.glb sets `mimeType` to `image/jpeg`, but the jpeg loader fails.

So they're PNGs...

**Comment by akien-mga** [comment]
The fix is good, I'd tweak the comment to be clearer and keep the jpg attempt in its scope (so that it's still covered by the comment):
```diff
diff --git a/modules/gltf/gltf_document.cpp b/modules/gltf/gltf_document.cpp
index 6594bba65e..68ef766d9b 100644
--- a/modules/gltf/gltf_document.cpp
+++ b/modules/gltf/gltf_document.cpp
@@ -3092,9 +3092,14 @@ Error GLTFDocument::_parse_images(Ref<GLTFState> state, const String &p_base_pat
                } else if (mimetype == "image/jpeg") { // Loader buffer as JPEG.
                        ERR_FAIL_COND_V(Image::_jpg_mem_loader_func == nullptr, ERR_UNAVAILABLE);
                        img = Image::_jpg_mem_loader_func(data_ptr, data_size);
-               } else {
+               }
+
+               if (img.is_null()) {
                        // We can land here if we got an URI with base64-encoded data with application/* MIME type,
-                       // and the optional mimeType property was not defined to tell us how to handle this data (or was invalid).
+                       // and the optional mimeType property was not defined to tell us how to handle this data.
+                       // This also lets us handle bogus mime types (e.g. `image/jpeg` but the data is actually a PNG).
+                       // That's not as per spec but bogus glb files do exist in production so we're lenient.
+
                        // So let's try PNG first, then JPEG.
                        ERR_FAIL_COND_V(Image::_png_mem_loader_func == nullptr, ERR_UNAVAILABLE);
                        img = Image::_png_mem_loader_func(data_ptr, data_size);
```

We could add a verbose print but that would require some further checks which are maybe not worth the trouble (e.g. `if (img.is_valid() && mimetype == "image/jpeg")` after the PNG load attempt, etc.).

File: modules/gltf/gltf_document.cpp [hunk 1]
```diff
@@ -3008,16 +3008,18 @@ Error GLTFDocument::_parse_images(Ref<GLTFState> state, const String &p_base_pat
...
 			img = Image::_jpg_mem_loader_func(data_ptr, data_size);
-		} else {
-			// We can land here if we got an URI with base64-encoded data with application/* MIME type,
-			// and the optional mimeType property was not defined to tell us how to handle this data (or was invalid).
-			// So let's try PNG first, then JPEG.
+		}
+        // We can land here if we got an URI with base64-encoded data with application/* MIME type,
+        // and the optional mimeType property was not defined to tell us how to handle this data.
+        // This also lets us handle bogus mime types (e.g. `image/jpeg` but the data is actually a PNG).
+        // That's not as per spec but bogus glb files do exist in production so we're lenient.
```
Comment (akien-mga) [inline modules/gltf/gltf_document.cpp:3015]: Wrong indentation, but if you put this comment in the main scope, you'll have to change the phrasing.

It's no longer a "we can land here" as all code goes through this scope. "here" referred to after `if (img.is_null())`.

**Comment by akien-mga** [comment]
So here's fixed documentation if you want to write the code like this:

```diff
diff --git a/modules/gltf/gltf_document.cpp b/modules/gltf/gltf_document.cpp
index 6594bba65e..3f839cd9fa 100644
--- a/modules/gltf/gltf_document.cpp
+++ b/modules/gltf/gltf_document.cpp
@@ -3086,24 +3086,32 @@ Error GLTFDocument::_parse_images(Ref<GLTFState> state, const String &p_base_pat
 
 		Ref<Image> img;
 
+		// First we honor the mime types if they were defined.
 		if (mimetype == "image/png") { // Load buffer as PNG.
 			ERR_FAIL_COND_V(Image::_png_mem_loader_func == nullptr, ERR_UNAVAILABLE);
 			img = Image::_png_mem_loader_func(data_ptr, data_size);
 		} else if (mimetype == "image/jpeg") { // Loader buffer as JPEG.
 			ERR_FAIL_COND_V(Image::_jpg_mem_loader_func == nullptr, ERR_UNAVAILABLE);
 			img = Image::_jpg_mem_loader_func(data_ptr, data_size);
-		} else {
-			// We can land here if we got an URI with base64-encoded data with application/* MIME type,
-			// and the optional mimeType property was not defined to tell us how to handle this data (or was invalid).
-			// So let's try PNG first, then JPEG.
+		}
+
+		// If we didn't pass the above tests, we attempt loading as PNG and then
+		// JPEG directly.
+		// This covers URIs with base64-encoded data with application/* type but
+		// no optional mimeType property, or bufferViews with a bogus mimeType
+		// (e.g. `image/jpeg` but the data is actually PNG).
+		// That's not *exactly* what the spec mandates but this lets us be
+		// lenient with bogus glb files which do exist in production.
+		if (img.is_null()) { // Try PNG first.
 			ERR_FAIL_COND_V(Image::_png_mem_loader_func == nullptr, ERR_UNAVAILABLE);
 			img = Image::_png_mem_loader_func(data_ptr, data_size);
-			if (img.is_null()) {
-				ERR_FAIL_COND_V(Image::_jpg_mem_loader_func == nullptr, ERR_UNAVAILABLE);
-				img = Image::_jpg_mem_loader_func(data_ptr, data_size);
-			}
+		}
+		if (img.is_null()) { // And then JPEG.
+			ERR_FAIL_COND_V(Image::_jpg_mem_loader_func == nullptr, ERR_UNAVAILABLE);
+			img = Image::_jpg_mem_loader_func(data_ptr, data_size);
 		}
 
+		// Now we've done our best, fix your scenes.
 		if (img.is_null()) {
 			ERR_PRINT(vformat("glTF: Couldn't load image index '%d' with its given mimetype: %s.", i, mimetype));
 			state->images.push_back(Ref<Texture>());

```

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Alright it's fine as is :)

**Comment by fire** [comment]
Ahhhhh,I merged your changes.

**Comment by akien-mga** [comment]
Thanks!

**Comment by akien-mga** [comment]
Cherry-picked for 3.4.

**Pull request merged**